titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java

```diff
@@ -0,0 +1,262 @@
+        Task task = taskEvent.getCurrentTask();
+
+        if (task.getStatus().getState() == TaskState.Started) {
+            handleTaskStartedEvent(task);
+        } else if (task.getStatus().getState() == TaskState.Finished) {
+            handleTaskFinishedEvent(task);
+        }
+    }
+
+    private void handleTaskStartedEvent(Task task) {
+        PlacementHistory placement = tryGetOrCreatePlacementHistory(task);
+        if (placement != null) {
+            placement.onTaskStarted();
+        }
+    }
+
+    private void handleTaskFinishedEvent(Task task) {
+        String reasonCode = task.getStatus().getReasonCode();
+        if (reasonCode == null) {
+            invariants.inconsistent("Task with status without reason code: taskId={}, status={}", task.getId(), task.getStatus());
+            return;
+        }
+
+        ErrorKind errorKind;
+        switch (reasonCode) {
+            case TaskStatus.REASON_NORMAL:
+            case TaskStatus.REASON_TASK_KILLED:
+            case TaskStatus.REASON_FAILED:
+            case TaskStatus.REASON_STUCK_IN_KILLING_STATE:
+            default:
+                return;
+            case TaskStatus.REASON_CRASHED:
+                errorKind = ErrorKind.ContainerCrash;
+                break;
+            case TaskStatus.REASON_TASK_LOST:
+                errorKind = isStartedTask(task) ? ErrorKind.LostAfterStarted : ErrorKind.LostBeforeStarted;
+                break;
+            case TaskStatus.REASON_STUCK_IN_STATE:
+            case TaskStatus.REASON_LOCAL_SYSTEM_ERROR:
+                errorKind = ErrorKind.LocalStartingError;
+                break;
+            case TaskStatus.REASON_TRANSIENT_SYSTEM_ERROR:
+            case TaskStatus.REASON_UNKNOWN_SYSTEM_ERROR:
+                errorKind = ErrorKind.SystemStartingError;
+                break;
+        }
+
+        PlacementHistory placement = tryGetOrCreatePlacementHistory(task);
+        if (placement != null) {
+            placement.onTaskFailure(errorKind);
+        }
+    }
+
+    private PlacementHistory tryGetOrCreatePlacementHistory(Task task) {
+        String hostId = task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_AGENT_HOST);
+        if (hostId == null) {
+            invariants.inconsistent("Task without host name assigned: taskId=%s", task.getId());
+            return null;
+        }
+
+        PlacementHistory placement = agentsPlacementHistory.get(hostId);
+        if (placement == null) {
+            String instanceId = task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_AGENT_INSTANCE_ID);
+            if (instanceId == null) {
+                invariants.inconsistent("Task without agent id assigned: taskId=%s", task.getId());
+                return null;
+            }
+            placement = agentManagementService.findAgentInstance(instanceId).map(PlacementHistory::new).orElse(null);
+        }
+
+        return placement;
+    }
+
+    private boolean isStartedTask(Task task) {
```

Comment:
I have seen these types of utilities a few times. Maybe we should start moving them to `JobFunctions` or another utility class